cmd/setup.go

```diff
@@ -0,0 +1,303 @@
+/*
+ * Minio Cloud Storage, (C) 2017 Minio, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package cmd
+
+import (
+	"fmt"
+	"net"
+	"strings"
+
+	"github.com/minio/minio-go/pkg/set"
+)
+
+// SetupType - enum for setup type.
+type SetupType int
+
+const (
+	// FSSetupType - FS setup type enum.
+	FSSetupType SetupType = 1
+
+	// XLSetupType - XL setup type enum.
+	XLSetupType = 2
+
+	// DistXLSetupType - Distributed XL setup type enum.
+	DistXLSetupType = 3
+)
+
+// Setup - interface denotes any minio setup.
+type Setup interface {
+	ServerAddr() string
+	Endpoints() EndpointList
+	Type() SetupType
+}
```

Comment:
> But what is the goal of doing these changes? is it to bring all the setup related global variables into a single global struct so as not to pollute the global name space?

This change will make globals removed and setup definition is encapsulated into one object and that will bring a context ability which will help unit tests have it own context to run (not in parallel).

> note that:
> 
>     serverAddr string
>     endPoints []Endpoint
> 
> does not seem to be useful once init is completed.

No. On every login to servers need to validate endpoints and associated `format.json` of remote servers.  endpoints are needed to be sent/received.  This is not done in the current code base.